modules/build/src/main/scala/scala/build/bsp/BspImpl.scala

```diff
@@ -282,70 +282,87 @@ final class BspImpl(
           def isJavaFile    = relPath.last.endsWith(".java")
           newOrDeletedFile && !isHidden && (isScalaFile || isJavaFile)
         }
-        val watcher0 = watcher.newWatcher()
+        val watcher0 = currentBloopSession.watcher.newWatcher()
         watcher0.register(elem.path.toNIO, Int.MaxValue)
         watcher0.addObserver {
           Build.onChangeBufferedObserver { event =>
             if (eventFilter(event))
-              watcher.schedule()
+              currentBloopSession.watcher.schedule()
           }
         }
       case _ =>
     }
 
-  val actualLocalClient = new BspClient(
+  private val actualLocalClient = new BspClient(
     threads.buildThreads.bloop.jsonrpc, // meh
     logger
   )
-  actualLocalClient.setProjectName(inputs.workspace, inputs.projectName, Scope.Main)
-  val localClient: b.BuildClient with BloopBuildClient =
+  private val localClient: b.BuildClient with BloopBuildClient =
     if (verbosity >= 3)
       new BspImpl.LoggingBspClient(actualLocalClient)
     else
       actualLocalClient
 
-  var remoteServer: BloopCompiler  = null
-  var actualLocalServer: BspServer = null
+  private def newBloopSession(inputs: Inputs): BloopSession = {
```

Comment:
Since this is essentially initializing a new session, I think it'd be better to make it a constructor/`apply` method for `BloopSession`. `BspImpl` is too big a source file already.

Reply:
I tried that [here](https://github.com/alexarchambault/scala-cli/commit/301d47b19b200e6288d4aeed82d2a3a765819b67), but I'm not too happy with the result: that method references a few others of `BspImpl`'s private methods, so we have to pass them as closures to `BloopSession.create` say, which is a bit ugly…

Reply:
eh, neither is ideal, I guess. ok then.